plutus-tx/src/PlutusTx/Builtins/Internal.hs

```diff
@@ -190,15 +211,20 @@ type family TwiddleBS fake where
              (C1 ('MetaCons "BuiltinByteString" x y)
                z)
 
--- We define a custom Generic instance because we want any generic code working
--- with `BuiltinByteString` to see it as a single-constructor type. In
--- particular, any serialization should be done without a constructor tag.
-instance Generic BuiltinByteString where
-  type Rep BuiltinByteString = TwiddleBS (Rep FakeBuiltinByteString)
-  {-# INLINE to #-}
-  to (M1 (M1 (M1 (K1 bs)))) = BuiltinByteString bs
-  {-# INLINE from #-}
-  from (BuiltinByteString bs) = M1 (M1 (M1 (K1 bs)))
+{- Note [Building a custom Generic instance]
+
+It's really easy to write custom `to` and `from` methods for `Generic`.  The
+trickier part is getting the `Rep` right. The `Rep` includes metadata about
+strictness, unpacking, constructor names, record selectors, etc. Some aspects
+are difficult or impossible to get right manually, particularly the exact
+package name and the "decided strictness". So instead of building the `Rep`
```

Comment:
You can get the "package name" (actually the unit-id) with `CURRENT_PACKAGE_KEY` CPP macro.
OTOH using the fake type with generic trick is quite clever :)

Reply:
@phadej, using a fake type is the only way I know to be sure the `DecidedStrictness` is right no matter how the module is compiled, and I think it's easier to see that the construction is correct than it would be to do it all by hand anyway. Copying the module name from the fake type makes it robust against module renaming too. For the sake of my curiosity, what is a `unit-id`?

Reply:
> what is a unit-id

It's a difficult question to answer exhaustively, but a short version unit-ids are the identifiers libraries (or more generally units) are registered with in package db (or more generally identified by GHC). Package id (which is package version + package number) is metadata used for humans. For ghc boot libraries unit-ids and package-ids are the same, for `cabal` installed packages (in store) they are different:

For example:

```
% head -n 10 /cabal/store/ghc-9.0.1/package.db/aeson-2.0.2.0-ceb356429ca3538b1da748cbb4d8187275f311ecac1e7bd3d68bd0e65c65d68a.conf
name:                 aeson
version:              2.0.2.0
visibility:           public
id:
    aeson-2.0.2.0-ceb356429ca3538b1da748cbb4d8187275f311ecac1e7bd3d68bd0e65c65d68a

key:
    aeson-2.0.2.0-ceb356429ca3538b1da748cbb4d8187275f311ecac1e7bd3d68bd0e65c65d68a

license:              BSD-3-Clause
```

(`key` was the predecessor of unit-ids before Backpack, IIRC, now there for compatibility reasons).